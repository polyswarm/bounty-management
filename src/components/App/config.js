const config = {
  url: process.env.REACT_APP_HOST,
};
export default config;
