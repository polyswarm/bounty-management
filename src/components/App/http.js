import BigNumber from 'bignumber.js';
class HttpApp {
  constructor(url, ws) {
    this.url = url;
    this.ws = ws;
  }

  getUnlockedWallet() {
    return fetch(this.url+'/accounts/active')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => !json.locked)
      .catch(() => false);
  }

  getWallets() {
    return fetch(this.url+'/accounts')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .catch(() => []);
  }

  getBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw Error('Cannot get bounties.');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(bounty => this.getAssertionsForBounty(bounty))
      .then(bountyAssertions => this.getArtifactsForBounty(bountyAssertions))
      .then(bounty => {
        const assertions = bounty.assertions;
        let files = bounty.artifacts;
        assertions.forEach((assertion) => {
          assertion.verdicts.forEach((verdict, index) => {
            const file = files[index];
            if (!verdict) {
              file.good++;
            }
            file.total++;
            file.assertions.push({
              author: assertion.author,
              bid: assertion.bid,
              verdict: verdict,
              metadata: assertion.metadata
            });
            files[index] = file;
          });
        });
        bounty.artifacts = files;
        return bounty;
      })
      .catch(() => null);
  }

  getArtifactsForBounty(bounty) {
    return fetch(this.url+'/artifacts/'+bounty.uri)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(files => {
        return files.map((file) => {
          return file.name;
        });
      })
      .then(filesnames => {
        return filesnames.map((name) => {
          const trimmed = name.trim();
          return {name: trimmed, good: 0, total: 0, assertions: []};
        });
      })
      .then(files => {
        bounty.artifacts = files;
        return bounty;
      });
  }

  getAssertionsForBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid+'/assertions')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(assertions => {
        return assertions.map((assertion) => {
          const bid = new BigNumber(assertion.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
          return {
            author: assertion.author,
            bid: bid,
            verdicts: assertion.verdicts,
            metadata: assertion.metadata,
          };
        });
      })
      .then(filtered => {
        bounty.assertions = filtered;
        return bounty;
      });
  }

  listenForAssertions(assertionAddedCallback) {
    // attach to websocket
    // anytime we get an assertion, check if it matches a guid
    // if it does, add it to the assertions for that object
    const ws = this.ws;
    const websocket = new WebSocket(ws);

    websocket.onmessage = (event) => {
      const message = JSON.parse(event.data);

      if (message.event === 'assertion') {
        const body = message.data;
        const bid = new BigNumber(body.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
        const assertion = {
          guid: body.bounty_guid,
          bid: bid,
          verdicts: body.verdicts,
          metadata: body.metadata,
          author: body.author,
        };
        assertionAddedCallback(assertion);
      }
    };
  }
}
export default HttpApp;
