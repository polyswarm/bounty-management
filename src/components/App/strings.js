const strings = {
  create: 'Create a Bounty',
  bountyError: 'Error: Failed to Post Bounty',
  requestAllData: 'Refreshing bounties',
  requestWallets: 'Updating Wallets',
  requestGetBounty: 'Getting Bounty'
};

export default strings;
