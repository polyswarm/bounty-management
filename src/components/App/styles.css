html,
body,
#root {
  height: 100%;
}

body {
  font-family: 'Roboto', sans-serif;
  margin: 0;
  overflow: hidden;
  padding: 0;
}

.BountyManager {
  color: white;
  height: 100%;
  left: 0;
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
}

.hex-background {
  background-color: #8500ff;
  background-image: url('/img/hexgrid_2x.png');
  background-size: 129px 149px;
}

.BountyManager-Intro {
  font-size: large;
}
