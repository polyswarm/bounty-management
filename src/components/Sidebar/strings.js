const strings = {
  logo: 'PolySwarm logo',
};
export default strings;
