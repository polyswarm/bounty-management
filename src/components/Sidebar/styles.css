.Sidebar {
  background-color: #8500ff;
  border-right: 2px solid #9c38ff;
  display: inline-block;
  height: 100%;
  left: 0;
  max-width: 300px;
  position: relative;
  top: 0;
}

.Sidebar-Content {
  color: white;
  height: 100%;
  overflow: auto;
}

.Sidebar-Header > img {
  box-sizing: border-box;
  height: inherit;
  left: 0;
  padding: 0 10px;
  width: 75%;
}

.Sidebar-Header {
  background-image: url('/img/hexgrid_2x.png');
  background-size: 129px 149px;
  border-bottom: 2px solid #9c38ff;
  display: block;
  height: 132px;
}
