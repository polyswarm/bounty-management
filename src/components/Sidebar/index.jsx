// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
// Bounty Imports
import ListItem from '../ListItem';
// Component Imports
import strings from './strings';
import './styles.css';

class Sidebar extends Component {
  constructor(props) {
    super(props);
    this.onSelectBounty = this.onSelectBounty.bind(this);
  }

  render() {
    const { props: { bounties, active } } = this;
    return (
      <div className='Sidebar'>
        <header className='Sidebar-Header'>
          <img src='/img/polyswarm.svg'
            alt={strings.logo}/>
        </header>
        <ul className='Sidebar-Content'>
          {bounties && bounties.length > 0 && (
            bounties.map((bounty, index) => {
              return(
                <ListItem className={`item-${index}`}
                  active={index === active}
                  alert={bounty.updated}
                  key={bounty.guid}
                  onClick={() => {this.onSelectBounty(index);}}
                  remove={() => {this.onBountyRemoved(index);}}>
                  {bounty.guid}
                </ListItem>
              );
            })
          )}
        </ul>
      </div>
    );
  }

  onSelectBounty(index) {
    const { props: { select } } = this;
    if (select) {
      select(index);
    }
  }

  onBountyRemoved(index) {
    const { props: { remove } } = this;
    if (remove) {
      remove(index);
    }
  }

}

Sidebar.defaultProps = {
  bounties: PropTypes.array.isRequired,
  select: PropTypes.func.isRequired,
  remove: PropTypes.func.isRequired,
};

export default Sidebar;
