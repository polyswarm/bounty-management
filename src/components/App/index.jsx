// Vendor imports
import React, { Component } from 'react';
// Bounty imports
import BountyCreate from '../BountyCreate';
import BountyInfo from '../BountyInfo';
import Sidebar from '../Sidebar';
import Header from '../Header';
import Welcome from '../Welcome';
// Component imports
import HttpApp from './http';
import config from './config';
import strings from './strings';
import './styles.css';

class App extends Component {
  constructor(props) {
    super(props);
    const {bounties, first} = this.preloadLocalStorage();
    this.http = new HttpApp(config.url, config.wsUrl);
    this.state = {
      isUnlocked: false,
      walletList: [],
      active: 0,
      bounties: bounties,
      create: false,
      first: first,
    };

    this.onAddBounty = this.onAddBounty.bind(this);
    this.onRemoveBounty = this.onRemoveBounty.bind(this);
    this.onSelectBounty = this.onSelectBounty.bind(this);
    this.onCreateBounty = this.onCreateBounty.bind(this);
    this.onCloseWelcome = this.onCloseWelcome.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.getData = this.getData.bind(this);
    this.getWallets = this.getWallets.bind(this);
  }

  componentDidUpdate(_, prevState) {
    const { state: { bounties } } = this;
    const { bounties: prevBounties } = prevState;
    const storageOutOfSync =  JSON.stringify(bounties) !== JSON.stringify(prevBounties);
    if(storageOutOfSync) {
      this.storeBounties(bounties);
    }
  }

  componentDidMount() {
    this.getWallets();
    this.getData();
  }

  render() {
    const {url} = config;
    const { state: { active, bounties, create, first, isUnlocked, walletList } } = this;

    return (
      <div className='App'>
        {first && (
          <Welcome onClick={this.onCloseWelcome}/>
        )}
        {!first && (
          <React.Fragment>
            <Sidebar bounties={bounties}
              active={active}
              remove={this.onRemoveBounty}
              select={this.onSelectBounty}/>
            <Header title={(bounties.length === 0 || create || active < 0) ? strings.create : bounties[active].guid}
              create={create}
              onClick={this.onCreateBounty}/>
            <div className='App-Content'>
              { (bounties.length === 0 || create || active < 0 ) && (
                <BountyCreate url={url}
                  isUnlocked={isUnlocked}
                  walletList={walletList}
                  onWalletChange={this.onWalletChangeHandler}
                  addBounty={this.onAddBounty}/>
              )}
              { !create && active >=0 && active < bounties.length && (
                <BountyInfo bounty={bounties[active]}/>
              )}
            </div>
          </React.Fragment>
        )}
      </div>
    );
  }

  onAddBounty(guid) {
    const bounties = this.state.bounties.slice();
    const bounty = {
      guid: guid,
      update: false,
      author: '',
      amount: '',
      artifactURI: '',
      expirationBlock: '',
      resolved: '',
      verdicts: '',
    };
    bounties.push(bounty);
    this.setState({bounties: bounties});
  }

  onCreateBounty() {
    this.setState({create: true, active: -1});
  }

  onCloseWelcome() {
    this.setState({first: false});
    this.markSeen();
  }

  onRemoveBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      bounties.splice(index, 1);
      this.setState({bounties: bounties});
    }
  }

  onSelectBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      const bounty = bounties[index];
      bounty.updated = false;
      bounties[index] = bounty;
      this.setState({active: index, create: false, bounties: bounties});
    }
  }

  onWalletChangeHandler(store) {
    this.setState({isUnlocked: store});
  }

  pullLocalBounties() {
    const bounties = this.state.bounties.slice();
    return bounties;
  };

  updateOnAssertion(assertion) {
    const bounties = this.state.bounties.slice();
    const guid = assertion.guid;
    const index = bounties.findIndex((bounty) => bounty.guid === guid);
    // let index = -1;
    // bounties.forEach((bounty, i) => {
    //   if (bounty.guid === guid) {
    //     index = i;
    //   }
    // });
    if (index >= 0) {
      const bounty = bounties[index];
      const a = {};
      a[assertion.author] = {
        bid: assertion.bid,
        metadata: assertion.metadata,
        verdicts: assertion.verdicts,
      };
      bounty.assertsions.push(a);
      const modified = bounty.artifacts.map((file, index) => {
        const f = file;
        if (!a.verdicts[index]) {
          f.good++;
        }
        f.total++;
        file.assertions.push({
          author: assertion.author,
          bid: assertion.bid,
          verdict: a.verdicts[index],
          metadata: assertion.metadata
        });
        return f;
      });
      bounty.artifacts = modified;
      bounty.update = true;
      bounties[index] = bounty;
    }
    this.setState({bounties: bounties});
  }

  getData() {
    const http = this.http;
    http.listenForAssertions(this.pullLocalBounties, this.updateOnAssertion);
    const bounties = this.state.bounties.slice();
    const promises = bounties.map((bounty) => {
      return http.getBounty(bounty)
        .then(b => {
          if (Object.keys(b) !== Object.keys(bounty)) {
            b.update = true;
          }
          return b;
        })
        .catch(() => bounty);
    });
    Promise.all(promises).then((values) => {
      // get updated state after download finishes
      const bounties = this.state.bounties.slice();
      values.forEach((value) => {
        const foundIndex = bounties.findIndex((bounty) => bounty.guid === value.guid);
        if (foundIndex >= 0) {
          bounties[foundIndex] = value;
        }
      });
      this.setState({bounties: bounties});
    });
  }

  getWallets() {
    const http = this.http;
    http.getWallets()
      .then(accounts => {
        this.setState({walletList: accounts});
      });

    http.getUnlockedWallet()
      .then((success) => this.setState({isUnlocked: success}));
  }

  storeBounties(bounties) {
    if (this.hasLocalStorage()) {
      localStorage.setItem('bounties', JSON.stringify(bounties));
    }
  }

  hasLocalStorage() {
    try {
      localStorage.setItem('x', 'y');
      localStorage.removeItem('x');
      return true;
    } catch(e) {
      return false;
    }
  }

  markSeen() {
    if (this.hasLocalStorage()) {
      localStorage.setItem('seen', JSON.stringify(true));
    }
  }

  preloadLocalStorage() {
    if (this.hasLocalStorage) {
      const bounties = JSON.parse(localStorage.getItem('bounties')) || [];
      const first = !JSON.parse(localStorage.getItem('seen'));
      return {bounties, first};
    } else {
      return {bounties: [], first: true};
    }
  }
}

export default App;
