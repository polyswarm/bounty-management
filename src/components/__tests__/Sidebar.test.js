import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson, mountToJson} from 'enzyme-to-json';
import Sidebar from '../Sidebar';
import App from '../App';

it('renders without crashing', () => {
  const bounties = [];
  const wrapper = render(<Sidebar bounties={bounties} />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('shows each bounty as a ListItem', () => {
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const wrapper = mount(<Sidebar bounties={bounties} />);
  expect(wrapper.find('.ListItem')).toHaveLength(2);
});

it('marks the selected row as .active',() => {
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const wrapper = mount(<Sidebar bounties={bounties} active={1}/>);

  expect(wrapper.find('.item-1').props().active).toBeTruthy();
  expect(wrapper.find('.item-0').props().active).toBeFalsy();
  expect(mountToJson(wrapper)).toMatchSnapshot();
});

it('sets the state when onSelectBounty called',() => {
  const select = jest.fn();
  const bounties = [{guid:'asdf'}, {guid:'demo'}];

  const wrapper = mount(<Sidebar select={select} bounties={bounties} />);
  const instance = wrapper.instance();

  instance.onSelectBounty(1);

  expect(select).toHaveBeenCalledWith(1);
});

it('selects the row that was clicked',() => {
  const select = jest.spyOn(App.prototype, 'onSelectBounty');
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const wrapper = mount(<App/>);
  wrapper.setState({first: false, bounties: bounties, active: 0});

  wrapper.find('.item-1').simulate('click');

  expect(select).toHaveBeenCalledWith(1);
  expect(wrapper.find('.item-1').props().active).toBeTruthy();
});

it('calls remove when onBountyRemoved called',() => {
  const remove = jest.fn();
  const bounties = [{guid:'asdf'}, {guid:'demo'}];
  const wrapper = mount(<Sidebar remove={remove} bounties={bounties} />);
  const instance = wrapper.instance();

  instance.onBountyRemoved(1);

  expect(remove).toHaveBeenCalledWith(1);
});

it('calls remove when RemoveButton clicked for a row',() => {
  const remove = jest.fn();
  const select = jest.fn();
  const bounties = [{guid:'asdf'}, {guid:'demo'}];

  const wrapper = mount(<Sidebar select={select} remove={remove} bounties={bounties} />);

  wrapper.find('.item-1').simulate('mouseEnter');
  wrapper.find('.item-1').find('.Remove-Button').simulate('click');

  expect(remove).toHaveBeenCalledWith(1);
  expect(select).toHaveBeenCalledTimes(0);
});

it('shows alert when bounty has updated: true', () => {
  const remove = jest.fn();
  const select = jest.fn();
  const bounties = [{guid:'asdf', updated: true}, {guid:'demo'}];

  const wrapper = mount(<Sidebar select={select} remove={remove} bounties={bounties} />);

  expect(wrapper.find('.item-0').find('.alert')).toHaveLength(1);
  expect(mountToJson(wrapper)).toMatchSnapshot();
});
