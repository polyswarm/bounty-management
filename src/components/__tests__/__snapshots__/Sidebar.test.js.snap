// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marks the selected row as .active 1`] = `
<Sidebar
  active={1}
  bounties={
    Array [
      Object {
        "guid": "asdf",
      },
      Object {
        "guid": "demo",
      },
    ]
  }
  remove={[Function]}
  select={[Function]}
>
  <div
    className="Sidebar"
  >
    <header
      className="Sidebar-Header"
    >
      <img
        alt="Polyswarm logo"
        src="/img/polyswarm.svg"
      />
    </header>
    <ul
      className="Sidebar-Content"
    >
      <ListItem
        active={false}
        className="item-0"
        key="asdf"
        onClick={[Function]}
        remove={[Function]}
      >
        <li
          className="ListItem"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div>
            <span
              className="ListItem-Child"
              onClick={[Function]}
            >
              asdf
            </span>
          </div>
        </li>
      </ListItem>
      <ListItem
        active={true}
        className="item-1"
        key="demo"
        onClick={[Function]}
        remove={[Function]}
      >
        <li
          className="ListItem active"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div>
            <span
              className="ListItem-Child"
              onClick={[Function]}
            >
              demo
            </span>
          </div>
        </li>
      </ListItem>
    </ul>
  </div>
</Sidebar>
`;

exports[`renders without crashing 1`] = `
<div
  class="Sidebar"
>
  <header
    class="Sidebar-Header"
  >
    <img
      alt="Polyswarm logo"
      src="/img/polyswarm.svg"
    >
      
    </img>
  </header>
  <ul
    class="Sidebar-Content"
  >
    
  </ul>
</div>
`;

exports[`shows alert when bounty has updated: true 1`] = `
<Sidebar
  bounties={
    Array [
      Object {
        "guid": "asdf",
        "updated": true,
      },
      Object {
        "guid": "demo",
      },
    ]
  }
  remove={[Function]}
  select={[Function]}
>
  <div
    className="Sidebar"
  >
    <header
      className="Sidebar-Header"
    >
      <img
        alt="Polyswarm logo"
        src="/img/polyswarm.svg"
      />
    </header>
    <ul
      className="Sidebar-Content"
    >
      <ListItem
        active={false}
        alert={true}
        className="item-0"
        key="asdf"
        onClick={[Function]}
        remove={[Function]}
      >
        <li
          className="ListItem"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div>
            <div
              className="alert"
            />
            <span
              className="ListItem-Child"
              onClick={[Function]}
            >
              asdf
            </span>
          </div>
        </li>
      </ListItem>
      <ListItem
        active={false}
        className="item-1"
        key="demo"
        onClick={[Function]}
        remove={[Function]}
      >
        <li
          className="ListItem"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div>
            <span
              className="ListItem-Child"
              onClick={[Function]}
            >
              demo
            </span>
          </div>
        </li>
      </ListItem>
    </ul>
  </div>
</Sidebar>
`;
